C.TXIO:PAYMENT:EQ_H_AMOUNT violated by TPCC#payment(TPCCContext, PaymentInput)

It seems that the `H_AMOUNT` value in the output and input are not equal. Full trace below.

```
Thread[fabric-txinvoke:5,5,main] 07:00:53:393 SEVERE  org.hyperledger.fabric.Logger error                                              Error during contract method executionorg.hyperledger.fabric.contract.ContractRuntimeException: Error during contract method execution
        at org.hyperledger.fabric.contract.execution.impl.ContractExecutionService.executeRequest(ContractExecutionService.java:78)
        at org.hyperledger.fabric.contract.ContractRouter.processRequest(ContractRouter.java:116)
        at org.hyperledger.fabric.contract.ContractRouter.invoke(ContractRouter.java:127)
        at org.hyperledger.fabric.shim.impl.ChaincodeInvocationTask.call(ChaincodeInvocationTask.java:106)
        at org.hyperledger.fabric.shim.impl.InvocationTaskManager.lambda$newTask$17(InvocationTaskManager.java:265)
        at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.base/java.lang.Thread.run(Unknown Source)
Caused by: org.jmlspecs.runtime.JmlAssertionError: /home/gradle/src/main/java/hu/bme/mit/ftsrg/chaincode/tpcc/TPCC.java:629: verify: JML postcondition is false
  private PaymentOutput payment(final TPCCContext ctx, final PaymentInput input)
                        ^
specs/hu/bme/mit/ftsrg/chaincode/tpcc/TPCC.jml:46: verify: Associated declaration: /home/gradle/src/main/java/hu/bme/mit/ftsrg/chaincode/tpcc/TPCC.java:629:
    @ ensures \result.getH_amount() == input.getH_amount(); // C.TXIO:PAYMENT:EQ_H_AMOUNT
      ^
        at org.jmlspecs.runtime.Utils.createException(Utils.java:142)
        at org.jmlspecs.runtime.Utils.assertionFailureL(Utils.java:94)
        at hu.bme.mit.ftsrg.chaincode.tpcc.TPCC.payment(TPCC.java:1)
        at hu.bme.mit.ftsrg.chaincode.tpcc.TPCC.payment(TPCC.java:135)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
        at java.base/java.lang.reflect.Method.invoke(Unknown Source)
        at org.hyperledger.fabric.contract.execution.impl.ContractExecutionService.executeRequest(ContractExecutionService.java:60)
        ... 8 more
.. caused by ../home/gradle/src/main/java/hu/bme/mit/ftsrg/chaincode/tpcc/TPCC.java:629: verify: JML postcondition is false
  private PaymentOutput payment(final TPCCContext ctx, final PaymentInput input)
                        ^
specs/hu/bme/mit/ftsrg/chaincode/tpcc/TPCC.jml:46: verify: Associated declaration: /home/gradle/src/main/java/hu/bme/mit/ftsrg/chaincode/tpcc/TPCC.java:629:
    @ ensures \result.getH_amount() == input.getH_amount(); // C.TXIO:PAYMENT:EQ_H_AMOUNT
      ^org.jmlspecs.runtime.JmlAssertionError: /home/gradle/src/main/java/hu/bme/mit/ftsrg/chaincode/tpcc/TPCC.java:629: verify: JML postcondition is false
  private PaymentOutput payment(final TPCCContext ctx, final PaymentInput input)
                        ^
specs/hu/bme/mit/ftsrg/chaincode/tpcc/TPCC.jml:46: verify: Associated declaration: /home/gradle/src/main/java/hu/bme/mit/ftsrg/chaincode/tpcc/TPCC.java:629:
    @ ensures \result.getH_amount() == input.getH_amount(); // C.TXIO:PAYMENT:EQ_H_AMOUNT
      ^
        at org.jmlspecs.runtime.Utils.createException(Utils.java:142)
        at org.jmlspecs.runtime.Utils.assertionFailureL(Utils.java:94)
        at hu.bme.mit.ftsrg.chaincode.tpcc.TPCC.payment(TPCC.java:1)
        at hu.bme.mit.ftsrg.chaincode.tpcc.TPCC.payment(TPCC.java:135)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
        at java.base/java.lang.reflect.Method.invoke(Unknown Source)
        at org.hyperledger.fabric.contract.execution.impl.ContractExecutionService.executeRequest(ContractExecutionService.java:60)
        at org.hyperledger.fabric.contract.ContractRouter.processRequest(ContractRouter.java:116)
        at org.hyperledger.fabric.contract.ContractRouter.invoke(ContractRouter.java:127)
        at org.hyperledger.fabric.shim.impl.ChaincodeInvocationTask.call(ChaincodeInvocationTask.java:106)
        at org.hyperledger.fabric.shim.impl.InvocationTaskManager.lambda$newTask$17(InvocationTaskManager.java:265)
        at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.base/java.lang.Thread.run(Unknown Source)
```

--- a/smart-contract/hyperledger-fabric/v2/java/build.gradle.kts
+++ b/smart-contract/hyperledger-fabric/v2/java/build.gradle.kts
@@ -80,7 +80,9 @@ tasks.withType<JavaCompile>().configureEach {
           else -> "rac"
         }
     options.isFork = true
-    options.compilerArgs.addAll(listOf("-jml", "-$mode", "-timeout", "30", "--nullable-by-default", "--specs-path", "specs/"))
+    options.compilerArgs.addAll(
+        listOf(
+            "-jml", "-$mode", "-timeout", "30", "--nullable-by-default", "--specs-path", "specs/"))
     options.forkOptions.javaHome = openJMLJavaHomeDir.asFile
   }
 }

--- a/smart-contract/hyperledger-fabric/v2/java/src/main/java/hu/bme/mit/ftsrg/chaincode/tpcc/TPCC.java
+++ b/smart-contract/hyperledger-fabric/v2/java/src/main/java/hu/bme/mit/ftsrg/chaincode/tpcc/TPCC.java
@@ -785,6 +785,7 @@ public final class TPCC implements ContractInterface {
             .fromWarehouse(warehouse)
             .fromCustomer(customer)
             .fromDistrict(district)
+            .fromHistory(history)
             .build();
     if (customer.getC_credit().equals("BC"))
       output.setC_data(customer.getC_data().substring(0, 200));

--- a/smart-contract/hyperledger-fabric/v2/java/src/main/java/hu/bme/mit/ftsrg/chaincode/tpcc/data/output/PaymentOutput.java
+++ b/smart-contract/hyperledger-fabric/v2/java/src/main/java/hu/bme/mit/ftsrg/chaincode/tpcc/data/output/PaymentOutput.java
@@ -5,6 +5,7 @@ package hu.bme.mit.ftsrg.chaincode.tpcc.data.output;
 import hu.bme.mit.ftsrg.chaincode.tpcc.TPCC;
 import hu.bme.mit.ftsrg.chaincode.tpcc.data.entity.Customer;
 import hu.bme.mit.ftsrg.chaincode.tpcc.data.entity.District;
+import hu.bme.mit.ftsrg.chaincode.tpcc.data.entity.History;
 import hu.bme.mit.ftsrg.chaincode.tpcc.data.entity.Warehouse;
 import hu.bme.mit.ftsrg.chaincode.tpcc.middleware.TPCCContext;
 import lombok.EqualsAndHashCode;
@@ -664,6 +665,12 @@ public final class PaymentOutput {
       this.c_credit_lim(customer.getC_credit_lim());
       this.c_discount(customer.getC_discount());
       this.c_balance(customer.getC_balance());
+      return this;
+    }
+
+    public DoPaymentOutputBuilder fromHistory(final History history) {
+      this.h_amount = history.getH_amount();
+      this.h_date = history.getH_date();
       return this;
     }
 

